[AMDGPU] Fix register class constraints for si-fold-operands pass when folding immediate into copies

mssefat opened the pull request:
Fixes https://github.com/llvm/llvm-project/issues/130020

This fixes an issue where the si-fold-operands pass would incorrectly fold immediate values into COPY instructions targeting av_32 registers.

The pass now checks register class constraints before attempting to fold the immediate.

mssefat committed:
[AMDGPU] Fix register class constraints for si-fold-operands pass

Fixes #130020

This fixes an issue where the si-fold-operands pass would incorrectly
fold immediate values into COPY instructions targeting av_32 registers,
which is illegal.

The pass now properly checks register class constraints before attempting to
fold the immediates.

mssefat commented:
@arsenm could you please review?

arsenm commented:
The tests for the issue are already in llvm/test/CodeGen/AMDGPU/fold-imm-copy.mir (and you should also mention this fixes issue number in the description)

mssefat commented:
@arsenm I updated accordingly. Please check.

arsenm commented on llvm/lib/Target/AMDGPU/SIFoldOperands.cpp:
```diff
@@ -1047,6 +1047,11 @@ void SIFoldOperandsImpl::foldOperand(
     if (MovOp == AMDGPU::COPY)
       return;
 
+    // Check if the destination register of the MOV operation belongs
+    // to a vector superclass. Folding would be illegal.
+    if (TRI->isVectorSuperClass(DestRC))
+      return;
```
Would be better to query the register class of the result register, and check if there is a common class between it and the virtual register class. In the future it would be better to replace the instruction and modify the register class to enable folds in some cases

mssefat replied:
Thanks for your suggestion. Based on your comment, I've implemented the check. The snippet looks like following:

```
    // Check if the destination register class has a common class
    // with the expected register class for the MOV instruction.
    const MCInstrDesc &MovDesc = TII->get(MovOp);
    if (MovDesc.operands()[0].RegClass != -1) {
      const TargetRegisterClass *ResRC = 
          TRI->getRegClass(MovDesc.operands()[0].RegClass);
      const TargetRegisterClass *CommonRC = 
          TRI->getCommonSubClass(DestRC, ResRC);

      if (!CommonRC) {
        return;
      }
    }
```

Can you please confirm if this aligns correctly with your suggestion or further adjustments are needed?

mssefat replied:
I have pushed the updated changes. Could you please review?

mssefat replied:
Ping

mssefat commented on llvm/test/CodeGen/AMDGPU/fold-imm-copy.mir:
```diff
@@ -185,52 +185,49 @@ body:             |
-
-# ...
+---
+name: s_mov_b32_inlineimm_copy_s_to_av_32
```
I have pushed the updated changes. Could you please review?

mssefat replied:
Ping

jayfoad commented on llvm/lib/Target/AMDGPU/SIFoldOperands.cpp:
```diff
@@ -1047,6 +1047,16 @@ void SIFoldOperandsImpl::foldOperand(
+    if (MovDesc.getNumDefs() > 0 && MovDesc.operands()[0].RegClass != -1) {
+      const TargetRegisterClass *ResRC = 
+          TRI->getRegClass(MovDesc.operands()[0].RegClass);
+      if (!DestRC -> hasSuperClassEq(ResRC)) return;
```
No space around `->`. Body of `if` on separate line.
```suggestion
      if (!DestRC->hasSuperClassEq(ResRC))
        return;
```

arsenm replied:
Can you use getCommonSubClass or getMatchingSuperRegClass instead of directly checking hasSuperClassEq

mssefat replied:
I have pushed the updated changes. Could you please review?

jayfoad commented on llvm/test/CodeGen/AMDGPU/fold-imm-copy.mir:
```diff
@@ -1,14 +1,23 @@
-
-# GCN-LABEL:       name: fold-imm-copy
-# GCN:             V_AND_B32_e32 65535
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
```
Please make a separate PR for this and commit it first. Then rebase the current PR.

mssefat replied:
Done and I have pushed the updated changes. Could you please review?

mssefat committed:
Removed changes for fold-imm-copy.mir and created a nfc for it with a separate PR

mssefat committed:
Fixes #130020

This fixes an issue where the si-fold-operands pass would incorrectly
fold immediate values into COPY instructions targeting av_32 registers,
which is illegal.

The pass now properly checks register class constraints before attempting to
fold the immediates.

mssefat committed:
[AMDGPU] Fix register class constraints for si-fold-operands pass

Fixes #130020

This fixes an issue where the si-fold-operands pass would incorrectly
fold immediate values into COPY instructions targeting av_32 registers,
which is illegal.

The pass now properly checks register class constraints before attempting to
fold the immediates.

mssefat commented on llvm/lib/Target/AMDGPU/SIFoldOperands.cpp:
```diff
@@ -1047,6 +1047,16 @@ void SIFoldOperandsImpl::foldOperand(
+    // is a superclass of (or equal to) the destination register class of the COPY (DestRC).
+    // If this condition fails, folding would be illegal.
+    const MCInstrDesc &MovDesc = TII->get(MovOp);
+    if (MovDesc.getNumDefs() > 0 && MovDesc.operands()[0].RegClass != -1) {
```
Done, can you please review?

arsenm commented on llvm/lib/Target/AMDGPU/SIFoldOperands.cpp:
```diff
@@ -1068,6 +1068,20 @@ void SIFoldOperandsImpl::foldOperand(
+        TRI->getRegClass(MovDesc.operands()[0].RegClass);
+    const TargetRegisterClass *CommonRC = TRI->getCommonSubClass(DestRC, ResRC);
+
+    if (!CommonRC || !DestRC->hasSuperClassEq(CommonRC))
+      return;
+
```
```suggestion
    if (!CommonRC)
      return;

```

hasSuperClassEq should be redundant with finding the common class

mssefat replied:
I think hasSuperClassEq is not redundant with finding common subclass. Consider this basic block where the copy operation is illegal:

bb.0:
  %1:av_32 = V_MOV_B32_e32 32, implicit $exec
  $agpr0 = COPY %1:av_32
  S_ENDPGM 0

In this example:

DestRC is AV_32
ResRC is VGPR_32
CommonRC is VGPR_32

If we only check for CommonRC, the check would allow folding in this case since CommonRC exists (VGPR_32). However, folding here is illegal. And I think we need the condition if ( !DestRC->hasSuperClassEq(CommonRC) return.

jayfoad replied:
Why do you need CommonRC? Isn't the whole check equivalent to `DestRC->hasSuperClassEq(ResRC)`?

jayfoad replied:
The code as you have written it is weird, because CommonRC is by construction a subclass of DestRC, and then you are checking that it is a superclass of DestRC. These can only both be true if CommonRC *is* DestRC.

mssefat replied:
They should be equivalent check. I am reverting to the previous use of DestRC->hasSuperClassEq(ResRC). @arsenm please let me know if I am missing something.

mssefat commented:
@arsenm @jayfoad Ping

mssefat committed:
Reverting to previous approach of using DestRC->hasSuperClassEq(ResRC)

mssefat committed:
Formatted SIFoldOperands.cpp
